flink-ml-lib/src/main/java/org/apache/flink/ml/common/datastream/AllReduceUtils.java

```diff
@@ -0,0 +1,286 @@
+                            receivedElement,
+                            bufStart,
+                            transBuf,
+                            0,
+                            getLengthOfChunk(startChunkId + chunk, receivedElement.length));
+                    out.collect(Tuple3.of(startChunkId + chunk, receivedElement.length, transBuf));
+                }
+            }
+        }
+    }
+
+    /** Aggregates partitioned array from other workers. */
+    private static class AllReduceSum
+            extends AbstractStreamOperator<Tuple4<Integer, Integer, Integer, double[]>>
+            implements OneInputStreamOperator<
+                            Tuple3<Integer, Integer, double[]>,
+                            Tuple4<Integer, Integer, Integer, double[]>>,
+                    BoundedOneInput {
+
+        /**
+         * A map that aggregates the received partial arrays. The key is chunkId, the value is
+         * (chunkId, totalLength, aggPartitionedArray).
+         */
+        Map<Integer, Tuple3<Integer, Integer, double[]>> sum = new HashMap<>();
+
+        @Override
+        public void endInput() {
+            int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+            for (Tuple3<Integer, Integer, double[]> ele : sum.values()) {
+                for (int taskId = 0; taskId < numTasks; taskId++) {
+                    int chunkId = ele.f0;
+                    int totalLength = ele.f1;
+                    double[] aggregatedSum = ele.f2;
+                    output.collect(
+                            new StreamRecord<>(
+                                    Tuple4.of(taskId, chunkId, totalLength, aggregatedSum)));
+                }
+            }
+        }
+
+        @Override
+        public void processElement(StreamRecord<Tuple3<Integer, Integer, double[]>> streamRecord) {
+            Tuple3<Integer, Integer, double[]> record = streamRecord.getValue();
+            int chunkId = record.f0;
+            int totalLength = record.f1;
+            double[] partialSum = record.f2;
+            if (sum.containsKey(chunkId)) {
+                if (sum.get(chunkId).f1 != totalLength) {
+                    throw new RuntimeException("The input double array must have same length.");
+                }
+                double[] cur = sum.get(chunkId).f2;
+                for (int i = 0; i < cur.length; i++) {
+                    cur[i] += partialSum[i];
+                }
+            } else {
+                sum.put(chunkId, Tuple3.of(chunkId, totalLength, partialSum));
+            }
+        }
+    }
+
+    /** Organizes the received chunks into a big array. */
+    private static class AllReduceRecv extends AbstractStreamOperator<double[]>
+            implements OneInputStreamOperator<
+                            Tuple4<Integer, Integer, Integer, double[]>, double[]>,
+                    BoundedOneInput {
+
+        /**
+         * used to store the reduced results. Note that if we use {@link
```

Comment:
nits: Start the comments with an upper-case letter, e.g. `Stores the reduced results`?

And does this comment means that this utility function can be used in the iteration ALL_ROUND mode? If so, do we still need to throw exception if there is more than one input?

Reply:
The comments are updated.

We still have to make sure that in each epoch, we only have one double array in each partition.